README.md
# Track

[![Test status](https://github.com/mlange-42/track/actions/workflows/tests.yml/badge.svg)](https://github.com/mlange-42/track/actions/workflows/tests.yml)
[![GitHub](https://img.shields.io/badge/github-repo-blue?logo=github)](https://github.com/mlange-42/track)
[![Docs](https://img.shields.io/static/v1?label=docs&message=mdBook&color=B7410E&logo=mdBook)](https://mlange-42.github.io/track/)
[![MIT license](https://img.shields.io/github/license/mlange-42/track)](https://github.com/mlange-42/track/blob/main/LICENSE)

Track is a time tracking command line tool.

*Track* is designed to be easy to use, and stores tracking entries in a human-readable plain-text format.

## Screenshots

![App Screenshot](./docs/src/images/document.png)

## Features

* Track your working time from the command line
* Natural language-like syntax
* Supports hierarchical project structure
* Records stored as plain-text files for human readability and editing
* Different types of text-based and graphical reports

## Getting started

See the **[:blue_book: Track Documentation](https://mlange-42.github.io/track/)** for a comprehensive user guide!

### Installation

**Using Go:**

```shell
go install github.com/mlange-42/track@latest
```

**Without Go:**

Download binaries for your OS from the [Releases](https://github.com/mlange-42/track/releases/).

### Usage

Any time tracking `track` is associated to a *Project*.
Before you can start tracking, create a project:

```shell
track create project MyProject
```

Now, start tracking time on the project:

```shell
track start MyProject
```

To stop tracking, use:

```shell
track stop
```

Get information about the current tracking status:

```shell
track status
```

See the [:blue_book: Track Documentation](https://mlange-42.github.io/track/) for details and more commands.

## References

* Heavily inspired by [`timetrace`](https://github.com/dominikbraun/timetrace) and [`klog`](https://github.com/jotaen/klog)


CHANGELOG.md
# Changelog

## [[unpublished]](https://github.com/mlange-42/track/compare/v0.3.7...main)

Nothing

## [[v0.3.7]](https://github.com/mlange-42/track/compare/v0.3.6...v0.3.7)

### Other

* Tweak record file format documentation (#138)
* More unit tests for the CLI commands (#139, #142)
* Internal refactoring of report renderers (#140)
* Added `staticcheck` to CI actions (#141)
* Fix all `staticcheck` warnings (#141)

## [[v0.3.6]](https://github.com/mlange-42/track/compare/v0.3.5...v0.3.6)

### Other

* Durations are printed with space padding (` 1:23`) instead of zero-padding (`01:23`) in reports (#132)
* Documented the record format (#134)
* More unit tests for packages `core` and `util`, coverage 65.4% (#135)
* Started adding unit tests for the CLI commands (#136)

## [[v0.3.5]](https://github.com/mlange-42/track/compare/v0.3.4...v0.3.5)

### Features

* Tags can have values, like `+tag=value` (#122)
* Command `report tags` shows a breakdown of tag values if a single tag is selected (#122)
* Projects can have required tags that must be given for each record, with a value (#125)

### Bugfixes

 * Fix note not shown on first record in day report (#121)
 * Show records in week and day report one cell shorter (#121)

## [[v0.3.4]](https://github.com/mlange-42/track/compare/v0.3.3...v0.3.4)

### Features

* Timeline reports now show the total time spent in addition to bars (#105)
* Timeline reports support CSV format via flag `--csv` (#105)
* Timeline reports in CSV format can have a column per project via flag `--table` (#105)

### Bugfixes

* Close file after saving project (#108)
* Fix retrieving record for status when a project is selected (#117)

### Other

* Added a lot of unit tests (#106, #108)
* Extend record and pause consistency checks (#108)
* Use goroutines for reading record files, to improve performance on large datasets (#112, #114)

## [[v0.3.3]](https://github.com/mlange-42/track/compare/v0.3.2...v0.3.3)

### Features

* New command `report tag` to list tags with time statistics (#93)
* Add flag `--dry` to `delete` command for dry-run (#96)
* New command `move project` to move projects between workspaces (#99)

### Bugfixes

* Fix #94 broken child project time aggregation (#95)

### Other

* More efficient filtering of records by explicit consideration of time (#97)
* Code cleanup without functional changes (#102)

## [[v0.3.2]](https://github.com/mlange-42/track/compare/v0.3.1...v0.3.2)

### Features

* Export records to JSON and YAML in addition to CSV (#88)
* Add total, work and pause durations to CSV export (#88)
* New command `report treemap` to generate an SVG treemap of time per project (#89)
* Flag `--rename` for command `edit project` for renaming projects (#90)
* Additional alias `?` for command `status` (#91)

## [[v0.3.1]](https://github.com/mlange-42/track/compare/v0.3.0...v0.3.1)

:warning: In case of an error after upgrading to `v0.3.1`, delete the config file `%USER%/.track/config.yml`.

### Features

* The data path (normally `%USER%/.track`) can be set via env var `TRACK_PATH` (#81)
* Configurable fill character for records in day report (#82)
* Adds flag `--7days` to week reports to show 7 days instead of a calendar week (#85)
* Adds flag `--copy` to the switch command to copy note and tags like in start command (#86)

### Bugfixes

* Fix crash when requesting the open record and there are no records at all (#80)
* Fix for different `yes` answers in confirm prompt (#84)

## [[v0.3.0]](https://github.com/mlange-42/track/compare/v0.2.1...v0.3.0)

### Features

* Switched to storing records in plain-text files with simple custom format (#52)
* Pause times can now be stored inside records instead of stop/start for a pause (#52)
* New command `pause` to pause a record, or insert a finished pause (#52)
* ~~The tag prefix is now `#` instead of `+` (record file comments are `//`) (#53)~~ (#62)
* Projects have a foreground color, in addition to the background color (#54)
* The `resume` command has flags `--at` and `--ago` to correct times (#55)
* The `start` command has a flag `--copy` to copy note and tags from the previous record in the same project (#58)
* The command `track edit record` can be used without arguments, or with only time as argument (#59)
* Command `track list tags` to list all tags, with number of occurrences (#60)
* Command `track edit day` to edit all records of a day in a single file (#61)
* Command `track report day` now prints a schedule-like report similar to `week` (#63)
* All `edit` subcommands have a flag `--dry` for dry-run (#68)
* All `edit` subcommands re-open the edited file if parsing/checks fail; improved error messages (#69)

### Bug fixes

* Apply `--at`/`--ago` also to start time in `switch`, not only to end time of previous record (#74)

### Other

* Removed command `break` (#52)
* Print only the first line of notes in the records list (#57)
* Record that starts the day before is included in `list records [DATE]` command (#62)
* Previous command `report day` (showing bar charts) renamed to `report chart` (#63)
* Print the latest record in `status`, in the default plain-text format (#75)
* Limit changes of end time when editing a single record (#76)
* Allow date words like "yesterday" in record selection for editing (#76)
* Extensive documentation under [mlange-42.github.io/track/](https://mlange-42.github.io/track/) (#71)

## [[v0.2.1]](https://github.com/mlange-42/track/compare/v0.2.0...v0.2.1)

### Bug fixes

* Limit time aggregation in reports to exact requested time range (#45)

## [[v0.2.0]](https://github.com/mlange-42/track/compare/v0.1.0...v0.2.0)

### Features

* Day report with horizontal timelines over projects as rows: `track report day [DATE]` (#25)
* Command `track break DURATION` to insert breaks (#28)
* Projects are structured in workspaces (#30)
* Records can be started and stopped with explicit time or offset from now (#31)
* Shows aggregated and self time of projects in report (#33)

cli/list.go
package cli

import (
	"fmt"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/gookit/color"
	"github.com/mlange-42/track/core"
	"github.com/mlange-42/track/out"
	"github.com/mlange-42/track/util"
	"github.com/spf13/cobra"
	"golang.org/x/exp/maps"
)

func listCommand(t *core.Track) *cobra.Command {
	list := &cobra.Command{
		Use:     "list",
		Short:   "List resources",
		Long:    "List resources",
		Aliases: []string{"l"},
		Run: func(cmd *cobra.Command, args []string) {
			_ = cmd.Help()
		},
	}

	list.AddCommand(listWorkspacesCommand(t))
	list.AddCommand(listProjectsCommand(t))
	list.AddCommand(listRecordsCommand(t))
	list.AddCommand(listColorsCommand(t))
	list.AddCommand(listTagsCommand(t))

	list.Long += "\n\n" + formatCmdTree(list)
	return list
}

func listProjectsCommand(t *core.Track) *cobra.Command {
	var includeArchived bool

	listProjects := &cobra.Command{
		Use:     "projects",
		Short:   "List all projects",
		Aliases: []string{"p"},
		Args:    util.WrappedArgs(cobra.NoArgs),
		RunE: func(cmd *cobra.Command, args []string) error {
			projects, err := t.LoadAllProjects()
			if err != nil {
				return fmt.Errorf("failed to list projects: %s", err)
			}
			if !includeArchived {
				pr := make(map[string]core.Project)
				for n, p := range projects {
					if !p.Archived {
						pr[n] = p
					}
				}
				projects = pr
			}

			var active string
			rec, err := t.OpenRecord()
			if err != nil {
				return fmt.Errorf("failed to list projects: %s", err)
			}
			if rec != nil {
				active = rec.Project
			}

			tree, err := t.ToProjectTree(projects)
			if err != nil {
				return fmt.Errorf("failed to list projects: %s", err)
			}
			formatter := util.NewTreeFormatter(
				func(t *core.ProjectNode, indent int) string {
					fillLen := 16 - (indent + utf8.RuneCountInString(t.Value.Name))
					name := t.Value.Name
					if fillLen < 0 {
						nameRunes := []rune(name)
						name = string(nameRunes[:len(nameRunes)+fillLen-1]) + "."
					}
					var str string
					if t.Value.Name == active {
						str = color.BgBlue.Sprintf("%s", name)
					} else {
						str = name
					}
					if fillLen > 0 {
						str += strings.Repeat(" ", fillLen)
					}
					str += " "
					str += t.Value.Render.Sprintf(" %s ", t.Value.Symbol)
					return str
				},
				2,
			)
			out.Print(formatter.FormatTree(tree))
			return nil
		},
	}
	listProjects.Flags().BoolVarP(&includeArchived, "archived", "a", false, "Include records from archived projects")

	return listProjects
}

func listWorkspacesCommand(t *core.Track) *cobra.Command {
	listWorkspaces := &cobra.Command{
		Use:     "workspaces",
		Short:   "List all workspaces",
		Aliases: []string{"w"},
		Args:    util.WrappedArgs(cobra.NoArgs),
		RunE: func(cmd *cobra.Command, args []string) error {
			ws, err := t.AllWorkspaces()
			if err != nil {
				return fmt.Errorf("failed to load workspaces: %s", err)
			}

			for i, w := range ws {
				if w == t.Workspace() {
					out.Print("%s", color.BgBlue.Sprintf("%s", w))
				} else {
					out.Print("%s", w)
				}
				if i < len(ws)-1 {
					out.Print("\n")
				}
			}
			return nil
		},
	}

	return listWorkspaces
}

func listRecordsCommand(t *core.Track) *cobra.Command {
	var includeArchived bool

	listProjects := &cobra.Command{
		Use:   "records [DATE]",
		Short: "List all records for a date",
		Long: `List all records for a date

The date can either be a date in default formatting, like "2022-12-31",
or a word like "yesterday" or  "today" (the default).`,
		Aliases:    []string{"r"},
		Args:       util.WrappedArgs(cobra.MaximumNArgs(1)),
		ArgAliases: []string{"date"},
		RunE: func(cmd *cobra.Command, args []string) error {
			date := util.ToDate(time.Now())
			var err error
			if len(args) > 0 {
				date, err = util.ParseDate(args[0])
				if err != nil {
					return fmt.Errorf("failed to load records: %s", err)
				}
			}

			records, err := t.LoadDateRecordsExact(date)
			if err != nil {
				if err == core.ErrNoRecords {
					out.Warn("no records for %s", date.Format(util.DateFormat))
					return nil
				}
				return fmt.Errorf("failed to load records: %s", err)
			}

			projects, err := t.LoadAllProjects()
			if err != nil {
				return fmt.Errorf("failed to export records: %s", err)
			}
			for _, record := range records {
				project := projects[record.Project]
				if includeArchived || !project.Archived {
					printRecord(record, project)
				}
			}
			return nil
		},
	}
	listProjects.Flags().BoolVarP(&includeArchived, "archived", "a", false, "Include records from archived projects")

	return listProjects
}

func listColorsCommand(t *core.Track) *cobra.Command {
	listColors := &cobra.Command{
		Use:     "colors",
		Short:   "Lists the 256 available colors",
		Aliases: []string{"c"},
		Args:    util.WrappedArgs(cobra.NoArgs),
		RunE: func(cmd *cobra.Command, args []string) error {
			printColorChart()
			return nil
		},
	}

	return listColors
}

func listTagsCommand(t *core.Track) *cobra.Command {
	var includeArchived bool

	listTags := &cobra.Command{
		Use:     "tags",
		Short:   "Lists all tags",
		Aliases: []string{"t"},
		Args:    util.WrappedArgs(cobra.NoArgs),
		RunE: func(cmd *cobra.Command, args []string) error {
			err := printTags(t, includeArchived)
			if err != nil {
				return fmt.Errorf("failed to list tags: %s", err.Error())
			}
			return nil
		},
	}
	listTags.Flags().BoolVarP(&includeArchived, "archived", "a", false, "Include records from archived projects")

	return listTags
}

func printRecord(r core.Record, project core.Project) {
	date := r.Start.Format(util.DateFormat)
	start := r.Start.Format(util.TimeFormat)

	var end string
	if r.HasEnded() {
		end = r.End.Format(util.TimeFormat)
	} else {
		end = util.NoTimeString
	}
	dur := r.Duration(util.NoTime, util.NoTime)
	pause := r.PauseDuration(util.NoTime, util.NoTime)

	fillLen := 16 - utf8.RuneCountInString(r.Project)
	name := r.Project
	if fillLen < 0 {
		nameRunes := []rune(name)
		name = string(nameRunes[:len(nameRunes)+fillLen-1]) + "."
	}

	fill := ""
	if fillLen > 0 {
		fill = strings.Repeat(" ", fillLen)
	}
	note := ""
	if r.Note != "" {
		parts := strings.SplitN(strings.ReplaceAll(r.Note, "\r\n", "\n"), "\n", 2)
		note = parts[0]
		if len(parts) > 1 {
			note += " ..."
		}
	}
	out.Print(
		"%s%s %s %s %s - %s (%5s + %5s)  %s\n", name, fill,
		project.Render.Sprintf(" %s ", project.Symbol),
		date, start, end, util.FormatDuration(dur, false), util.FormatDuration(pause, false),
		note,
	)
}

func printColorChart() {
	var row, block, i uint8

	out.Print("%s", color.C256(0, true).Sprintf("%3d", 0))
	out.Print(" ")
	for i = 1; i < 16; i++ {
		out.Print("%s", color.S256(0, i).Sprintf("%3d", i))
		out.Print(" ")
	}
	out.Print("\n\n")

	const rowOffset uint8 = 6
	const blockOffset uint8 = 36

	for _, start := range []uint8{16, 124} {
		for row = 0; row < 6; row++ {
			for block = 0; block < 3; block++ {
				idx := start + row*rowOffset + block*blockOffset
				for i = 0; i < 6; i++ {
					if row < 3 {
						out.Print("%s", color.C256(idx+i, true).Sprintf("%3d", idx+i))
					} else {
						out.Print("%s", color.S256(0, idx+i).Sprintf("%3d", idx+i))
					}
					out.Print(" ")
				}
				out.Print("  ")
			}
			out.Print("\n")
		}
		out.Print("\n")
	}

	for i = 232; i < 244; i++ {
		out.Print("%s", color.C256(i, true).Sprintf("%3d", i))
		out.Print(" ")
	}
	out.Print("\n")
	for i := 244; i <= 255; i++ {
		out.Print("%s", color.S256(0, uint8(i)).Sprintf("%3d", i))
		out.Print(" ")
	}
}

func printTags(t *core.Track, includeArchived bool) error {
	projects, err := t.LoadAllProjects()
	if err != nil {
		return err
	}

	tags := map[string]int{}
	values := map[string]map[string]bool{}

	filters := []core.FilterFunction{}
	if !includeArchived {
		filters = append(filters, core.FilterByArchived(false, projects))
	}

	fn, results, _ := t.AllRecordsFiltered(core.NewFilter(filters, util.NoTime, util.NoTime), false)

	go fn()
	for res := range results {
		if res.Err != nil {
			return res.Err
		}
		for tag, value := range res.Record.Tags {
			if v, ok := tags[tag]; ok {
				tags[tag] = v + 1
				values[tag][value] = true
			} else {
				tags[tag] = 1
				values[tag] = map[string]bool{value: true}
			}
		}
	}

	keys := maps.Keys(tags)
	sort.Strings(keys)

	for _, tag := range keys {
		v := maps.Keys(values[tag])
		sort.Strings(v)
		out.Print("%16s %4d", tag, tags[tag])
		if len(v) > 1 || (len(v) > 0 && v[0] != "") {
			out.Print(" [%s]", strings.Join(v, " "))
		}
		out.Print("\n")
	}

	return nil
}
